// astro/dist/build/page.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {enumerable: true, configurable: true, writable: true, value}) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
import path from "path";
import {compile as compilePathToRegexp} from "path-to-regexp";
import {fileURLToPath} from "url";
import {validateCollectionModule, validateCollectionResult} from "../util.js";
import {generateRSS} from "./rss.js";
function getPageType(filepath) {
  if (/\$[^.]+.astro$/.test(filepath.pathname))
    return "collection";
  return "static";
}
async function buildCollectionPage({astroConfig, filepath, astroRuntime, snowpackRuntime, site, buildState}) {
  const {pages: pagesRoot} = astroConfig;
  const srcURL = filepath.pathname.replace(pagesRoot.pathname, "");
  const pagesPath = astroConfig.pages.pathname.replace(astroConfig.projectRoot.pathname, "");
  const snowpackURL = `/_astro/${pagesPath}${srcURL}.js`;
  const mod = await snowpackRuntime.importModule(snowpackURL);
  validateCollectionModule(mod, filepath.pathname);
  const pageCollection = await mod.exports.createCollection();
  validateCollectionResult(pageCollection, filepath.pathname);
  let {route, paths: getPaths = () => [{params: {}}]} = pageCollection;
  const toPath = compilePathToRegexp(route);
  const allPaths = getPaths();
  const allRoutes = allPaths.map((p) => toPath(p.params));
  const builtURLs = new Set();
  async function loadPage(url) {
    if (builtURLs.has(url)) {
      return;
    }
    builtURLs.add(url);
    const result = await astroRuntime.load(url);
    if (result.statusCode === 200) {
      const outPath = path.posix.join(url, "/index.html");
      buildState[outPath] = {
        srcPath: filepath,
        contents: result.contents,
        contentType: "text/html",
        encoding: "utf8"
      };
    }
    return {url, result};
  }
  const loadResults = await Promise.all(allRoutes.map(loadPage));
  for (const loadResult of loadResults) {
    if (!loadResult) {
      continue;
    }
    const result = loadResult.result;
    if (result.statusCode >= 500) {
      throw new Error(result.error);
    }
    if (result.statusCode === 200) {
      const {collectionInfo} = result;
      if (collectionInfo == null ? void 0 : collectionInfo.rss) {
        if (!site) {
          throw new Error(`[${srcURL}] createCollection() tried to generate RSS but "buildOptions.site" missing in astro.config.mjs`);
        }
        const feedURL = "/feed" + loadResult.url + ".xml";
        const rss = generateRSS(__spreadProps(__spreadValues({}, collectionInfo.rss), {site}), {srcFile: srcURL, feedURL});
        buildState[feedURL] = {
          srcPath: filepath,
          contents: rss,
          contentType: "application/rss+xml",
          encoding: "utf8"
        };
      }
      if (collectionInfo == null ? void 0 : collectionInfo.additionalURLs) {
        await Promise.all([...collectionInfo.additionalURLs].map(loadPage));
      }
    }
  }
}
async function buildStaticPage({astroConfig, buildState, filepath, astroRuntime}) {
  const {pages: pagesRoot} = astroConfig;
  const url = filepath.pathname.replace(pagesRoot.pathname, "/").replace(/.(astro|md)$/, "").replace(/\/index$/, "/");
  const result = await astroRuntime.load(url);
  if (result.statusCode !== 200) {
    let err = result.error;
    if (!(err instanceof Error))
      err = new Error(err);
    err.filename = fileURLToPath(filepath);
    throw err;
  }
  const outFile = path.posix.join(url, "/index.html");
  buildState[outFile] = {
    srcPath: filepath,
    contents: result.contents,
    contentType: "text/html",
    encoding: "utf8"
  };
}
export {
  buildCollectionPage,
  buildStaticPage,
  getPageType
};
