// astro/dist/build/stats.js
import {info, table} from "../logger.js";
import {underline, bold} from "kleur/colors";
import gzipSize from "gzip-size";
function createURLStats() {
  return new Map();
}
function createBundleStats() {
  return new Map();
}
async function addBundleStats(bundleStatsMap, code, filename) {
  const gzsize = await gzipSize(code);
  bundleStatsMap.set(filename, {
    size: Buffer.byteLength(code),
    gzipSize: gzsize
  });
}
function mapBundleStatsToURLStats({urlStats, depTree, bundleStats}) {
  var _a;
  for (let [srcPath, stats] of bundleStats) {
    for (let url of urlStats.keys()) {
      if (depTree[url] && depTree[url].js.has("/" + srcPath)) {
        (_a = urlStats.get(url)) == null ? void 0 : _a.stats.push(stats);
      }
    }
  }
}
async function collectBundleStats(buildState, depTree) {
  const urlStats = createURLStats();
  await Promise.all(Object.keys(buildState).map(async (id) => {
    if (!depTree[id])
      return;
    const stats = await Promise.all([...depTree[id].js, ...depTree[id].css, ...depTree[id].images].map(async (url) => {
      if (!buildState[url])
        return void 0;
      const stat = {
        size: Buffer.byteLength(buildState[url].contents),
        gzipSize: await gzipSize(buildState[url].contents)
      };
      return stat;
    }));
    urlStats.set(id, {
      dynamicImports: new Set(),
      stats: stats.filter((s) => !!s)
    });
  }));
  return urlStats;
}
function logURLStats(logging, urlStats) {
  const builtURLs = [...urlStats.keys()].map((url) => url.replace(/index\.html$/, ""));
  builtURLs.sort((a, b) => a.localeCompare(b, "en", {numeric: true}));
  info(logging, null, "");
  const log = table(logging, [60, 20]);
  log(info, "   " + bold(underline("Pages")), bold(underline("Page Weight (GZip)")));
  const lastIndex = builtURLs.length - 1;
  builtURLs.forEach((url, index) => {
    var _a;
    const sep = index === 0 ? "\u250C" : index === lastIndex ? "\u2514" : "\u251C";
    const urlPart = " " + sep + " " + url;
    const bytes = ((_a = urlStats.get(url) || urlStats.get(url + "index.html")) == null ? void 0 : _a.stats.map((s) => s.gzipSize).reduce((a, b) => a + b, 0)) || 0;
    const kb = (bytes * 1e-3).toFixed(2);
    const sizePart = kb + " kB";
    log(info, urlPart + "index.html", sizePart);
  });
}
export {
  addBundleStats,
  collectBundleStats,
  createBundleStats,
  createURLStats,
  logURLStats,
  mapBundleStatsToURLStats
};
