// astro/dist/build/bundle/js.js
import {fileURLToPath} from "url";
import {rollup} from "rollup";
import {terser} from "rollup-plugin-terser";
import {createBundleStats, addBundleStats} from "../stats.js";
function collectJSImports(buildState) {
  const imports = new Set();
  for (const id of Object.keys(buildState)) {
    if (buildState[id].contentType === "application/javascript")
      imports.add(id);
  }
  return imports;
}
async function bundleJS(imports, {runtime, dist}) {
  const ROOT = "astro:root";
  const root = `
  ${[...imports].map((url) => `import '${url}';`).join("\n")}
`;
  const inputOptions = {
    input: [...imports],
    plugins: [
      {
        name: "astro:build",
        resolveId(source, imported) {
          if (source === ROOT) {
            return source;
          }
          if (source.startsWith("/")) {
            return source;
          }
          if (imported) {
            const outUrl = new URL(source, "http://example.com" + imported);
            return outUrl.pathname;
          }
          return null;
        },
        async load(id) {
          if (id === ROOT) {
            return root;
          }
          const result = await runtime.load(id);
          if (result.statusCode !== 200) {
            return null;
          }
          return result.contents.toString("utf-8");
        }
      }
    ]
  };
  const build = await rollup(inputOptions);
  const outputOptions = {
    dir: fileURLToPath(dist),
    format: "esm",
    exports: "named",
    entryFileNames(chunk) {
      const {facadeModuleId} = chunk;
      if (!facadeModuleId)
        throw new Error(`facadeModuleId missing: ${chunk.name}`);
      return facadeModuleId.substr(1);
    },
    plugins: [
      terser()
    ]
  };
  const stats = createBundleStats();
  const {output} = await build.write(outputOptions);
  await Promise.all(output.map(async (chunk) => {
    const code = chunk.code || "";
    await addBundleStats(stats, code, chunk.fileName);
  }));
  return stats;
}
export {
  bundleJS,
  collectJSImports
};
