// astro/dist/search.js
import {existsSync} from "fs";
import path from "path";
import {fileURLToPath} from "url";
import glob from "tiny-glob";
function findAnyPage(candidates, astroConfig) {
  for (let candidate of candidates) {
    const url = new URL(`./${candidate}`, astroConfig.pages);
    if (existsSync(url)) {
      const pagesPath = astroConfig.pages.pathname.replace(astroConfig.projectRoot.pathname, "");
      return {
        fileURL: url,
        snowpackURL: `/_astro/${pagesPath}${candidate}.js`
      };
    }
  }
  return false;
}
async function searchForPage(url, astroConfig) {
  const reqPath = decodeURI(url.pathname);
  const base = reqPath.substr(1);
  if (reqPath.endsWith("/")) {
    const candidates2 = [`${base}index.astro`, `${base}index.md`];
    const location2 = findAnyPage(candidates2, astroConfig);
    if (location2) {
      return {
        statusCode: 200,
        location: location2,
        pathname: reqPath
      };
    }
  } else {
    const candidates2 = [`${base}.astro`, `${base}.md`];
    let location2 = findAnyPage(candidates2, astroConfig);
    if (location2) {
      return {
        statusCode: 200,
        location: location2,
        pathname: reqPath
      };
    }
  }
  const candidates = [`${base}/index.astro`, `${base}/index.md`];
  const location = findAnyPage(candidates, astroConfig);
  if (location) {
    return {
      statusCode: 301,
      location: null,
      pathname: reqPath + "/"
    };
  }
  const hasExt = !!path.extname(reqPath);
  if (!location && !hasExt) {
    const collectionLocation = await findCollectionPage(reqPath, astroConfig);
    if (collectionLocation) {
      return {
        statusCode: 200,
        location: collectionLocation,
        pathname: reqPath
      };
    }
  }
  if (reqPath === "/500") {
    return {
      statusCode: 200,
      location: {
        fileURL: new URL("./frontend/500.astro", import.meta.url),
        snowpackURL: `/_astro_frontend/500.astro.js`
      },
      pathname: reqPath
    };
  }
  return {
    statusCode: 404
  };
}
async function findCollectionPage(reqPath, astroConfig) {
  const cwd = fileURLToPath(astroConfig.pages);
  const allCollections = {};
  const files = await glob("**/$*.{astro,md}", {cwd, filesOnly: true});
  for (const srcURL of files) {
    const pagesPath = astroConfig.pages.pathname.replace(astroConfig.projectRoot.pathname, "");
    const snowpackURL = `/_astro/${pagesPath}${srcURL}.js`;
    const reqURL = "/" + srcURL.replace(/\.(astro|md)$/, "").replace(/(^|[\/])\$/, "$1").replace(/index$/, "");
    allCollections[reqURL] = {snowpackURL, fileURL: new URL(srcURL, astroConfig.pages)};
  }
  let collectionMatchState = reqPath;
  do {
    if (allCollections[collectionMatchState]) {
      return allCollections[collectionMatchState];
    }
    collectionMatchState = collectionMatchState.substring(0, collectionMatchState.lastIndexOf("/"));
  } while (collectionMatchState.length > 0);
}
export {
  searchForPage
};
